Allow parallel testing, take 2 - (was #158)

Babar opened the pull request:
Basically a re-do of #158.
This got rolled back because @babar wrote 3 different PRs and one of them was bad.
This one wasn't. Technically it wasn't good either, as it only did `test`, not really `untest`, `run` or `keeptesting`.

This re-do fixes this.

Still the same features:
* Can parametrize the number of threads in the config, or on the command line, default to the number of cores
* Reports hosts depending whether they failed ssh or something else
* Doesn't have any extra dependency, as the thread pool is done in a very simple way

Test Plan:
Ran it on my machine. Tried `test` and `untest`, both worked.

Signed-off-by: Olivier Raginel <babar@fb.com>

Babar committed:
Allow parallel testing, take #2

Summary:
Basically a re-do of #158.
This got rolled back because @babar wrote 3 different PRs and one of them was bad.
This one wasn't. Technically it wasn't good either, as it only did `test`, not really `untest`, `run` or `keeptesting`.

This re-do fixes this.

Still the same features:
* Can parametrize the number of threads in the config, or on the command line, default to the number of cores
* Reports hosts depending whether they failed ssh or something else
* Doesn't have any extra dependency, as the thread pool is done in a very simple way

Test Plan:
Ran it on my machine. Tried test and untest, both worked.

Reviewers: nreeves, dcavalca, jaymzh

Subscribers:

Tasks:

Tags:
Signed-off-by: Olivier Raginel <babar@fb.com>

jaymzh commented:
IIRC, this broke internal wrappers... so CC @NaomiReeves on that.

But you definitely should make sure you've tested this with internal wrappers.

Babar commented:
The other PR broke the wrappers, from what I remember. The part where I moved the ARGS to be hostnames, 'cause some parts of our wrappers rely on TT just dropping unknown parameters.
I've fixed the wrapper too, but this is mostly to get speed improvements when testing a bunch of hosts.
Also, I've fixed the lint, so updating the PR shortly.

Babar committed:
Fix the lint and remove debug messages

Babar committed:
Prefix runchef IO stream with the hostname

Babar commented:
Oh one thing I'd love feedback on:
* How do people think we should deal with failures?
* How do we deal with logs?
In `test`, we report the failures and exit with:
* 0 if no failure
* 1 if all hosts failed and at least one failure was due to ssh connection issue
* 2 otherwise (meaning at least one host failed for whatever reason and at least one succeeded)
* 3 if all hosts failed because they were already tested by another user
I kept this logic for the other modes (so `untest`, `run` and `keeptesting`), even though 3 cannot happen.

As for logging, `run` can get very noisy with all the threads writing at the same time, and not knowing what's what. So I started writing something to prefix the logs with the hostname. But then I went down the rabbit hole, so I'll try to finish this some other night.

gogsbread commented on lib/taste_tester/commands.rb:
```diff
@@ -63,6 +63,69 @@ def self.status
+        # Poor man thread pool manager: keeping it simple
+        nb_threads_over_max = host_threads.length - TasteTester::Config.parallel_hosts
+        if nb_threads_over_max >= 0
+          host_threads[nb_threads_over_max].join
```
shouldn't this `join` be rescued as well?

Babar replied:
Good point. I guess this never happened in my testing 'cause I have a gazillion of cores :)

jaymzh commented:
> Oh one thing I'd love feedback on:
> 
> * How do people think we should deal with failures?
> * How do we deal with logs?
>   In `test`, we report the failures and exit with:
> * 0 if no failure
> * 1 if all hosts failed and at least one failure was due to ssh connection issue
> * 2 otherwise (meaning at least one host failed for whatever reason and at least one succeeded)
> * 3 if all hosts failed because they were already tested by another user
>   I kept this logic for the other modes (so `untest`, `run` and `keeptesting`), even though 3 cannot happen.
> 
> As for logging, `run` can get very noisy with all the threads writing at the same time, and not knowing what's what. So I started writing something to prefix the logs with the hostname. But then I went down the rabbit hole, so I'll try to finish this some other night.

I'm glad you asked!

A lot of this is actually the kind of stuff that we (we = TM folks, not FB folks, so past-past-past-life) solved over in https://github.com/txcketmaster-xx/onall - it forks a bunch of ssh's, joins 'em all back, and handles output nicely. It has a few options - live, buffered, prefixed-by-hostname, in-a-directory one-file-per-host, etc. and I think it's a great model for how to manage output and make it usable for folks.

As for return values: 0 - no errors, we can all agree on that. But maybe some more nuance for errors:

* All failed, all SSH failures (no test-setup/run failures)
* All failed, all because in-testing
* All failed, other test-setup/run failures
* All failed, mix of issues
* Some failed, all SSH related
* Some failed, all because in-testing
* Some failed, other test-setup/run failures
* Some failed, mix of issues

Further, I might recommend outputing a JSON structure of host=>failure, perhaps on stderr, or perhaps to a configurable file.

Babar commented:
Stupid question, but what's the difference between `onall` and `parallel`? They're bot written in perl, and seem to be doing pretty much the same thing. I mean, `parallel --tag ssh {} ...`. Haven't read the entire code yet, but I know `parallel`'s code fairly well, which is why I'm asking.

Babar committed:
Parallelize cookbook upload too

Summary:
As upload can take a very long time, we can just have a dedicated thread for it

Commit:
Fix all the lints

Commit:
Handle the exceptions better, and also handle the threads we join when we run out of threads in our poor man thread pool implementation

Commit:
Enhance reporting when things go wrong, including when hosts takes a while to finish

dafyddcrosby commented on lib/taste_tester/host.rb:
```diff
@@ -26,6 +26,29 @@
 
 module TasteTester
+  # Wrapper around IO to stream, prefixing the lines with the hostname
+  class IO < IO
```
Makes this a little easier to understand
```suggestion
  class IO < ::IO
```

jaymzh replied:
David is right - this looks like your subclassing yourself, we should be clear.

Babar committed:
Merge branch 'main' of https://github.com/facebook/taste-tester into parallel-testing-v2

Babar committed:
Remove debug line

jaymzh commented on lib/taste_tester/commands.rb:
```diff
@@ -63,7 +63,94 @@ def self.status
+          thread_status = _join_and_handle_ssh_exception(host_thread, 5)
+          unless thread_status.nil?
+            long_running_threads.delete host_thread
+            return_code.merge! [thread_status].to_h
```
```suggestion
            return_code.merge!([thread_status].to_h)
```

jaymzh commented on lib/taste_tester/commands.rb:
```diff
@@ -63,7 +63,94 @@ def self.status
+        running_threads += 1
+      end
+
+      hosts.to_set.each do |hostname|
```
we can't start these until the upload is done.

Babar replied:
Why? There is no point in waiting for the upload. You can't run it, but setting a host to taste-test has almost no relation with upload. The only caveat is that you need to ensure you have a working chef-zero instance, so you know the port number. But upload itself can totally be done in parallel.

jaymzh replied:
See other longer comment

jaymzh commented on lib/taste_tester/host.rb:
```diff
@@ -50,16 +73,9 @@ def runchef
       transport << TasteTester::Config.chef_client_command
 
-      io = IO.new(1)
+      io = TasteTester::IO.new(name, 1)
```
magic numbers are bad. What is 1?

Babar replied:
You do realize I didn't change that, right? I just added the name as a parameter :) I also have no idea what `1` means here.

jaymzh replied:
I do, but you touch the line you get to fix it. :)

Babar commented on lib/taste_tester/commands.rb:
```diff
@@ -80,9 +167,8 @@ def self.test
           logger.warn('Ignoring --linkonly because --really not set')
         end
-        upload
+        do_upload = true
```
I disagree. I initially did what you said, but then realized upload was taking longer and longer, so made a thread for it. Why do you think it has to complete before we set hosts to testing?

jaymzh replied:
Because it's entirely possible someone sees web12345 is done in the output, and goes and runs Chef on it. But in fact, that may not be safe to do for a long time. Remember that we're not necessarily starting with a blank chef-zero where the chef run will break early. It could be an existing one which month-old configs and you could revert that machine (or machines!) to a completely non-working state. A pretty common workflow would be that I parallel test 10 hosts, and those 10 hosts are machines I"m already SSH'd to because I"m preping the test. As I see the names go by from `tt`, I run chef. You're now expecting people to have an unreasonable level of understanding of internals to know that's not safe.

I feel pretty strongly this is the wrong design.

Even for,  autotest or something like it where the automation understands and is going to wait for all threads anyway, I can see wanting to save some time here, so let's discuss that possibility. Let's say we really want this and we hide it behind a flag like --unsafe-parallel-upload and make the help very clear no hosts can have chef run until TT exits successfully.  If we do that - what happens when every host succeeds but the upload fails? How do we indicate that? Or, what about when some hosts succeed, some fail, and the upload fails. Now we report partial success... except, that's not true either, even the "successful" hosts cannot safely have Chef run on them (see aforementioned old chef-zero state). Worse, we're now in a super inconsistent state where a bunch of hosts are pointed do a Chef server with a completely unknown state that requires either strong guarantees about the parent system cleaning them up in a timely fashion, or manual intervention from a human post-haste.

All this comes down to the design principal in the current system we don't touch a production host until _after_ we're sure we're pointing it at what the user expects. We don't point it at unknown stuff and hope for the best, and this code does just that, which is incredibly dangerous.

Babar commented on taste_tester.gemspec:
```diff
@@ -14,7 +14,7 @@
 Gem::Specification.new do |s|
   s.name = 'taste_tester'
-  s.version = '0.0.19'
+  s.version = '0.0.21'
```
Honestly, I just like 21 :) We totally should change that.

Babar commented on lib/taste_tester/commands.rb:
```diff
@@ -63,7 +63,95 @@ def self.status
+      return hostname, host_thread[:status]
+    rescue TasteTester::Exceptions::AlreadyTestingError => e
+      logger.error("User #{e.username} is already testing on #{hostname}")
+      return hostname, 42
```
IIRC, I used the same constants as the current implementation. But I agree it's better to use actual constants.

Babar commented on lib/taste_tester/commands.rb:
```diff
@@ -63,7 +63,94 @@ def self.status
+      end
+
+      print_remaining = true
+      while !long_running_threads.empty?
```
```suggestion
      until long_running_threads.empty?
```
